tsfc/kernel_interface.py

```diff
@@ -0,0 +1,467 @@
+from __future__ import absolute_import
+
+import numpy
+
+import coffee.base as coffee
+
+from tsfc import gem
+from tsfc.fiatinterface import create_element
+from tsfc.mixedelement import MixedElement
+from tsfc.node import traversal
+from tsfc.coffee import SCALAR_TYPE
+
+
+class Kernel(object):
+    __slots__ = ("ast", "integral_type", "oriented", "subdomain_id",
+                 "coefficient_numbers", "__weakref__")
+    """A compiled Kernel object.
+
+    :kwarg ast: The COFFEE ast for the kernel.
+    :kwarg integral_type: The type of integral.
+    :kwarg oriented: Does the kernel require cell_orientations.
+    :kwarg subdomain_id: What is the subdomain id for this kernel.
+    :kwarg coefficient_numbers: A list of which coefficients from the
+        form the kernel needs.
+    """
+    def __init__(self, ast=None, integral_type=None, oriented=False,
+                 subdomain_id=None, coefficient_numbers=()):
+        # Defaults
+        self.ast = ast
+        self.integral_type = integral_type
+        self.oriented = oriented
+        self.subdomain_id = subdomain_id
+        self.coefficient_numbers = coefficient_numbers
+        super(Kernel, self).__init__()
+
+
+class KernelBuilderBase(object):
+    """Helper class for building local assembly kernels."""
+
+    def __init__(self, interior_facet=False):
+        """Initialise a kernel builder.
+
+        :arg interior_facet: kernel accesses two cells
+        """
+        assert isinstance(interior_facet, bool)
+        self.interior_facet = interior_facet
+
+        self.prepare = []
+        self.finalise = []
+
+        self.coefficient_map = {}
+
+    def apply_glue(self, prepare=None, finalise=None):
+        """Append glue code for operations that are not handled in the
+        GEM abstraction.
+
+        Current uses: mixed interior facet mess
+
+        :arg prepare: code snippets to be prepended to the kernel
+        :arg finalise: code snippets to be appended to the kernel
+        """
+        if prepare is not None:
+            self.prepare.extend(prepare)
+        if finalise is not None:
+            self.finalise.extend(finalise)
+
+    def construct_kernel(self, name, args, body):
+        """Construct a COFFEE function declaration with the
+        accumulated glue code.
+
+        :arg name: function name
+        :arg args: function argument list
+        :arg body: function body (:class:`coffee.Block` node)
+        :returns: :class:`coffee.FunDecl` object
+        """
+        body.open_scope = False
+        body_ = coffee.Block(self.prepare + [body] + self.finalise)
```

Comment:
Now that I officially assume that `body` is a `coffee.Block`, I guess I can do just that. Maybe I can even add an assertion.